### DISCUSSION TITLE
MariaDB all my databases have disappeared

### DISCUSSION THREAD

**Discussion by murraycollingwood** [Homebrew/discussions#4042] [Getting started]
I have just upgraded my Mac to Ventura.  As part of this upgrade I had to re-brew a few items, including MariaDB.  
After the re-brew all of my databases have disappeared and only the default databases are now visible.
Is there a backup of the previous databases?  
Or are they still there but invisible to me?

**Comment by SMillerDev** [comment]
I'm not sure what a re-brew is. But what does `brew ls | grep mariadb` say?

**Reply by murraycollingwood** [reply]
Re-brew = When mariadb is already installed, but I run this anyway: brew install mariadb 

```
murraycollingwood@Jacaranda % brew ls | grep mariadb
mariadb

```
I can run mysql and when I show databases I get:
```
murraycollingwood@Jacaranda trunk % mysql -uroot -p
Enter password: 
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 66
Server version: 10.10.2-MariaDB Homebrew

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
5 rows in set (0.009 sec)

MariaDB [(none)]> 

```
There should be another 10-12 databases here... there was yesterday!
What happened?

**Reply by SMillerDev** [reply]
What I would expect is that you were on an older version of mariadb and your reinstall now means that it's reading from a different directory. What does `ls -al $(brew --prefix)/var/mysql` say?

**Reply by murraycollingwood** [reply]
```
murraycollingwood@Jacaranda trunk % ls -al $(brew --prefix)/var/mysql
total 246856
drwxr-xr-x   17 murraycollingwood  admin        544 13 Dec 07:45 .
drwxrwxr-x    8 murraycollingwood  admin        256 13 Dec 06:51 ..
-rw-rw----    1 murraycollingwood  admin      64814 13 Dec 21:13 Jacaranda.local.err
-rw-rw----    1 murraycollingwood  admin          6 13 Dec 07:45 Jacaranda.local.pid
-rw-rw----    1 murraycollingwood  admin     417792 13 Dec 07:55 aria_log.00000001
-rw-rw----    1 murraycollingwood  admin         52 13 Dec 07:45 aria_log_control
-rw-rw----    1 murraycollingwood  admin          9 13 Dec 07:45 ddl_recovery.log
-rw-r-----    1 murraycollingwood  admin        946 13 Dec 07:45 ib_buffer_pool
-rw-rw----    1 murraycollingwood  admin  100663296 13 Dec 07:48 ib_logfile0
-rw-rw----    1 murraycollingwood  admin   12582912 13 Dec 06:49 ibdata1
-rw-rw----    1 murraycollingwood  admin   12582912 13 Dec 07:45 ibtmp1
-rw-rw----    1 murraycollingwood  admin          0 13 Dec 07:45 multi-master.info
drwx------   90 murraycollingwood  admin       2880 13 Dec 06:49 mysql
-rw-r--r--    1 murraycollingwood  admin         15 13 Dec 06:49 mysql_upgrade_info
drwx------    3 murraycollingwood  admin         96 13 Dec 06:49 performance_schema
drwx------  106 murraycollingwood  admin       3392 13 Dec 06:49 sys
drwx------    3 murraycollingwood  admin         96 13 Dec 06:49 test
```

Okay - my databases are all 'innodb' type databases, so they should be in the 'ibdata1' file.
12.5 MB of data is the default allocation for a fresh install.  This is not my data...

**Reply by murraycollingwood** [reply]
I did some more hunting and found this:
```
murraycollingwood@Jacaranda mysql-5.7.22-macos10.13-x86_64 % sudo ls -al data
Password:
total 364080
drwxr-x---   22 _mysql  _mysql       704 13 Dec 07:45 .
drwxr-xr-x   13 root    wheel        416 30 Dec  2018 ..
-rw-r-----    1 _mysql  _mysql        56  2 May  2018 auto.cnf
drwxr-x---  107 _mysql  _mysql      3424 11 Aug  2020 crm
drwxr-x---   57 _mysql  _mysql      1824  2 May  2018 focus_capext
drwxr-x---  192 _mysql  _mysql      6144  2 May  2018 focus_capraw
drwxr-x---  483 _mysql  _mysql     15456  2 May  2018 focus_caps
drwxr-x---   12 _mysql  _mysql       384  2 May  2018 focus_focus
drwxr-x---   21 _mysql  _mysql       672  2 May  2018 focus_sabre
drwxr-x---  647 _mysql  _mysql     20704  8 Dec 14:15 focus_sobs
-rw-r-----    1 _mysql  _mysql      2801 13 Dec 07:45 ib_buffer_pool
-rw-r-----    1 _mysql  _mysql  50331648 13 Dec 07:45 ib_logfile0
-rw-r-----    1 _mysql  _mysql  50331648 16 Nov 16:21 ib_logfile1
-rw-r-----    1 _mysql  _mysql  79691776 13 Dec 07:45 ibdata1
-rw-r-----    1 _mysql  _mysql      9318  2 May  2018 kohekohe.local.err
drwxr-x---   77 _mysql  _mysql      2464  2 May  2018 mysql
-rw-r-----    1 _mysql  _mysql   5348047 13 Dec 07:45 mysqld.local.err
drwxr-x---   90 _mysql  _mysql      2880  2 May  2018 performance_schema
drwxr-x---   91 _mysql  _mysql      2912 16 Dec  2019 phree
drwxr-x---   91 _mysql  _mysql      2912 13 Nov  2019 phreeOld
drwxr-x---   91 _mysql  _mysql      2912 25 Oct  2019 phreenz
drwxr-x---  108 _mysql  _mysql      3456  2 May  2018 sys
murraycollingwood@Jacaranda mysql-5.7.22-macos10.13-x86_64 % pwd
/usr/local/mysql-5.7.22-macos10.13-x86_64
```

So what steps should I go through to recover this data?

My guess would be:

1. brew uninstall mariadb (to remove the current version)
2. brew install mariadb@???
3. Copy this directory back to /usr/local/var/mysql
4. mysqldump the data
5. brew uninstall mariadb (to remove the old version)
6. brew install mariadb (to install the latest version)
7. mysql to load the data dumped in step 4.

How does that sound?

**Reply by SMillerDev** [reply]
It looks by the naming there like you never had mariadb before and it was just MySQL instead. If that is the case I would suggest installing mysql@5.7 and running it using `mysqld_safe --datadir=/usr/local/mysql-5.7.22-macos10.13-x86_64` to see if your data is still accessible.

**Reply by murraycollingwood** [reply]
I'm struggling to make anything work here.
I can't install mysql@5.7, I can't do much with brew atm.
```
murraycollingwood@Jacaranda ~ % brew upgrade          
Error: Cannot install in Homebrew on ARM processor in Intel default prefix (/usr/local)!
Please create a new installation in /opt/homebrew using one of the
"Alternative Installs" from:
  https://docs.brew.sh/Installation
You can migrate your previously installed formula list with:
  brew bundle dump
```

I have no idea how to fix this.  
Homebrew is installed and was working, so why is it now complaining?

**Reply by SMillerDev** [reply]
> Homebrew is installed and was working, so why is it now complaining?

Because you're no longer calling it through Rosetta so it doesn't know what to do with the intel based install.

**Reply by murraycollingwood** [reply]
This has been a nightmare to try and work through, complicated somewhat by the change in architecture (x86 -> m1)
Just to recap, I'm running OSX Venture 13.1.  It's a recent upgrade that ditched my previous mysql@5.7, here's my current version information at the time of writing:
```
murraycollingwood@Jacaranda ~ % mysql --version
mysql  Ver 15.1 Distrib 10.10.2-MariaDB, for osx10.18 (arm64) using  EditLine wrapper
```

I couldn't find anything to shutdown the mariadb instance I had running, in the end I resorted to kill -9.  Not nice, but necessary it seems.

I was then able to install mysql@5.7, and again struggled to find commands to get this to run the daemon.  (mysql@5.7 installs a plugin in the 'System settings', use this to start and stop the mysql database)  Suddenly all of my old databases and data came back into view - hurrah!

Used mysqldump to unload each database.

Shutdown the mysql@5.7 but I couldn't find any method for uninstalling this.  Every time I tried brew uninstall mysql@5.7 it kept telling me there was no keg.  I tried preceding this with the arch... but nothing worked.  As far as I know it's probably still sitting on my disk somewhere, but it is supposedly harmless when not in use.  I deleted the plugin from 'System settings', so maybe that will stop it from being run.

Now I've installed mariadb: brew install mariadb
The post install didn't help - it should have run some initial configuration but didn't.  How to get this to run in the background, I thought this should happen automatically, but it didn't.  Finally found a mysqld_safe command which seems to have worked.

Next, I couldn't find any password capable of logging in.  I couldn't set any password to log in.  It was installed but I was locked out.

Went through the process of shutting down (kill -9 again because there doesn't appear to be any other way to shutdown mariadb that I was able to find, and I tried lots of options I found in google).  Deleted the database directory, and then ran the brew post install for mariadb.  This was supposed to reset any passwords that were set... it didn't.  It was still installed and I was still locked out.  Finally used my OSX username, without a password, and was able to access mysql.  I couldn't find any documentation about this, and frankly it's incredibly insecure to leave this so open, even on a development machine.  The post install should prompt for an initial password to be used with my username.

So now I have had to set my own password, create some additional users, (which I can't see, "select * from mysql.user;" no longer works, and there is no helpful information anywhere that I could find.  Checking the database "information_schema" I found a table called "user_privileges" which appeared to contain the privilege information.  No idea where the user account information is kept.

Next, create some databases (to restore the data into), but nothing worked.  
```
MariaDB [information_schema]> create database test;
ERROR 1006 (HY000): Can't create database 'test' (errno: 2 "No such file or directory")
```

This looks like a permission error, and I'm sure it is.  Everything I read told me to check the ownership of the daemon process, and the ownership of the 'datadir'.  
The daemon process has inherited my username, running under my OSX username.
The datadir also has my OSX username for ownership.
PS I didn't change any of these from their defaults - it would seem the install added them this way.
I'm not against all of this running under my username, but the problem is that it isn't working. 

I had a quick check, there is a user "_mysql" defined, and a group called "wheel".
Are these the ones I need to apply, or are there new ones for 'mariadb' that I'm not aware of?
Please, any help in fixing this permission problem???

**Reply by p-linnane** [reply]
If MySQL was installed on Intel (`/usr/local/`), you will need to run `arch -x86_64 brew uninstall mysql@5.7`. Anything from your old Intel-based install needs to be run with Rosetta 2 (`arch -x86_64`) to interact with it. Commands you run normally will affect your Apple Silicon-based install in `/opt/homebrew/`

**Reply by murraycollingwood** [reply]
```
murraycollingwood@Jacaranda etc % arch -x86_64 brew uninstall mysql@5.7
Error: No such keg: /opt/homebrew/Cellar/mysql@5.7
```
There's nothing like mysql* in that folder.
However, over in /usr/local/Cellar
```
murraycollingwood@Jacaranda etc % ls -al /usr/local/Cellar    
...miss some lines here...
drwxr-xr-x   3 murraycollingwood  wheel    96 15 Dec 22:02 mysql@5.7
...more lines here...
```

I couldn't find any method for uninstalling this one.

**Reply by p-linnane** [reply]
Do you have Rosetta installed? Running `arch -x86_64` should let you interact with the old install (in `/usr/local/`).

**Reply by murraycollingwood** [reply]
Yes - I had to use it to get mysql@5.7 installed.

**Reply by p-linnane** [reply]
How about `/usr/local/bin/brew uninstall mysql@5.7`?

**Reply by murraycollingwood** [reply]
YES!  That worked, thank you.

**Reply by murraycollingwood** [reply]
I think the permissions (using my username) are working now.
I had to stop and start mariadb using: 
```
brew services restart mariadb
```
